CONSOLE-922: Support AppliedClusterResourceQuota for normal users

rebeccaalpert opened the pull request:
# List Pages

I added logic so the resource link knows what the current namespace is. We need to pass the user's current namespace to the resource link if it's an ACRQ because the ACRQ itself doesn't supply namespace metadata.

Scoped namespace:
<img width="1078" alt="Screen Shot 2021-09-24 at 5 33 14 PM" src="https://user-images.githubusercontent.com/7014965/134742491-45597258-8fb6-494f-bb0c-ca28775e0035.png">

All namespaces:
<img width="1072" alt="Screen Shot 2021-09-24 at 5 34 45 PM" src="https://user-images.githubusercontent.com/7014965/134742506-8cb083ab-fc29-4c7f-b579-a25fb08fb3cd.png">

Support for search page (links work):
<img width="1422" alt="Screen Shot 2021-09-28 at 1 14 30 PM" src="https://user-images.githubusercontent.com/7014965/135134565-2f38450c-e0f2-4e95-91c4-3bba2b0396e3.png">

Support for API explorer page (instances links work):
<img width="1427" alt="Screen Shot 2021-09-28 at 1 15 14 PM" src="https://user-images.githubusercontent.com/7014965/135134524-42c539de-2e3b-4904-bd5f-230b8684099c.png">

# Details Page

New details page:
<img width="1096" alt="Screen Shot 2021-09-28 at 4 13 56 PM" src="https://user-images.githubusercontent.com/7014965/136033181-7494787b-5947-4080-8976-11290e7c8321.png">

New total used field (total used is cluster scope and used is scoped to the current namespace):
<img width="1095" alt="Screen Shot 2021-09-28 at 1 07 06 PM" src="https://user-images.githubusercontent.com/7014965/135133301-43534ce7-2173-429c-9102-e502084fe4e4.png">

We show ACRQs in the Resource Quota card in the project overview:
<img width="549" alt="Screen Shot 2021-10-04 at 2 13 08 PM" src="https://user-images.githubusercontent.com/7014965/136032517-48a39bcd-b194-4e54-a66b-6c935841f65c.png">
<img width="544" alt="Screen Shot 2021-10-04 at 2 13 15 PM" src="https://user-images.githubusercontent.com/7014965/136032551-1b4ba746-5515-40a9-b684-defe54a9eb70.png">

I'm basing the meat of the list page part of the PR on some work Sam did a few years ago: https://github.com/spadgett/console/commits/applied-cluster-quota. 

Fixes https://issues.redhat.com/browse/CONSOLE-922.

# How to test:

1. Add new user (we are calling this user `rebecca` here) and login
2. Create a new project: `oc new-project ralpert-test`
3. Log in as admin
4. Create a new role for them: `oc create role rqviewer --verb=get,list,watch --resource=resourcequota,appliedclusterresourcequota -n ralpert-test`
5. Assign the role: `oc adm policy add-role-to-user rqviewer rebecca --role-namespace=ralpert-test`
6. Create a cluster resource quota:
```
oc create clusterquota for-user \
     --project-annotation-selector openshift.io/requester=rebecca \
     --hard pods=10 \
     --hard secrets=20
```
7. Change user to `rebecca` and run `oc get appliedclusterresourcequota`. If you can see an ACRQ, you should see it in the UI list.
8. The link should take you to a custom ACRQ details page.

rebeccaalpert commented:
/test frontend

spadgett commented on frontend/packages/console-shared/src/constants/common.ts:
```diff
@@ -65,6 +65,7 @@ export enum FLAGS {
   CAN_LIST_NODE = 'CAN_LIST_NODE',
   CAN_LIST_PV = 'CAN_LIST_PV',
   CAN_LIST_CRD = 'CAN_LIST_CRD',
+  CAN_LIST_CLUSTER_QUOTA = 'CAN_LIST_CLUSTER_QUOTA',
```
Instead of adding a flag, I would just do an access review in the component. The `CAN_LIST_FOO` flags are typically needed to hide nav items, but not for general access reviews.

spadgett commented on frontend/public/components/default-resource.tsx:
```diff
@@ -102,22 +104,29 @@ export const DefaultList: React.FC<TableProps & { kinds: string[] }> = (props) =
     const menuActions = [...Kebab.getExtensionsActionsForKind(kindObj(kind)), ...common];
+    const appliedClusterQuotaReference = referenceForModel(AppliedClusterResourceQuotaModel);
+    const resourceNamespace =
+      kind === appliedClusterQuotaReference ? namespace : obj.metadata.namespace;
```
We shouldn't be changing the default resource table. Add a AppliedClusterResourceQuota specific table instead.

spadgett commented on frontend/public/components/factory/table.tsx:
```diff
@@ -270,6 +270,7 @@ const VirtualBody: React.FC<VirtualBodyProps> = (props) => {
     scrollTop,
     width,
     getRowProps,
+    namespace,
```
We should find a way to implement this without needing to pass through a new `namespace` prop. if this needed if you write your own AppliedClusterResource table instead of using the default table?

rebeccaalpert replied:
The actual ResourceQuota page doesn't use the DefaultList component like the Search and API Explorer pages. The changes in table.tsx were specifically for the actual ResourceQuota page. There wasn't a more straightforward way that I could see to get the namespace to the ResourceQuotaTableRow. It doesn't have access to the `match` URL information.

rebeccaalpert replied:
I'm totally open to suggestions; I'm just not seeing anything we can do otherwise unless we're able to somehow feed the namespace to the ResourceQuotaTableRow through Redux or local storage.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -42,10 +47,18 @@ const clusterResourceQuotaMenuActions = [
+    return resourceQuotaReference;
+  }
+
+  return canListClusterQuota ? clusterQuotaReference : appliedClusterQuotaReference;
```
So I think we probably always want to use `AppliedClusterResourceQuota` regardless of access. That way we can differentiate between used for project and total used, and we can even skip the access check.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,17 +70,36 @@ export const getQuotaResourceTypes = (quota) => {
+    const currentNamespace = sessionStorage.getItem(LAST_NAMESPACE_NAME_LOCAL_STORAGE_KEY);
+    used = used.filter((ns) => ns.namespace === currentNamespace);
+    used = _.get(used[0], ['status', 'used', resourceType]);
+    totalUsedPath = ['status', 'total', 'used'];
+    totalUsed = _.get(quota, [...totalUsedPath, resourceType]);
```
Why not just...?

```suggestion
    totalUsed = _.get(quota, ['status', 'total', 'used', resourceType]);
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,17 +70,36 @@ export const getQuotaResourceTypes = (quota) => {
+  // "used" is the data for the current namespace and "totalUsed" is the cluster-scoped data
+  if (isACRQ) {
+    used = _.get(quota, ['status', 'namespaces']);
+    const currentNamespace = sessionStorage.getItem(LAST_NAMESPACE_NAME_LOCAL_STORAGE_KEY);
```
This isn't how you want to get the current namespace. It should be part of the URL.

We also need to make sure we're handling "All Projects."

rebeccaalpert commented:
With suggested "always show ACRQ behavior":

**Admin:**
<img width="1105" alt="Screen Shot 2021-09-30 at 1 13 58 PM" src="https://user-images.githubusercontent.com/7014965/135501091-9b3a1e74-6f3b-46bd-89b3-4cfc8401a89e.png">

**Regular user:**
<img width="1104" alt="Screen Shot 2021-09-30 at 1 15 33 PM" src="https://user-images.githubusercontent.com/7014965/135501162-e76aa8e9-d8b5-45db-96d9-2feff8e288b8.png">

I addressed PR feedback where possible. @spadgett - I moved the ACRQ logic in the table row out into its own table row; let me know if you meant something else in your comment. Please also let me know if you have thoughts on the table component if you're against passing the namespace down.

I also added a test for the ACRQ ResourceUsageRow.

spadgett commented:
> I addressed PR feedback where possible. @spadgett - I moved the ACRQ logic in the table row out into its own table row; let me know if you meant something else in your comment. Please also let me know if you have thoughts on the table component if you're against passing the namespace down.

We definitely don't want to pass `namespace` down as a prop. Don't tables have a `customData` prop you can use?

An alternative that might be better is transitioning to the new table API: https://github.com/openshift/console/pull/8829

spadgett commented:
@rebeccaalpert We should update the project overview page as well to show cluster quotas.

rebeccaalpert commented:
All set @spadgett. I added ACRQs to the project dashboard.

<img width="272" alt="Screen Shot 2021-10-01 at 3 35 11 PM" src="https://user-images.githubusercontent.com/7014965/135677082-d7401dd9-e2e9-4068-81a6-cb98298e0db8.png">

spadgett commented:
@rebeccaalpert Sorry, I wasn't clear. I was thinking of the `ResourceQuotaCard` on the project dashboard:

https://github.com/openshift/console/blob/22c6951efe7c4bca87f3f934063b9f4dcb0a4058/frontend/public/components/dashboard/project-dashboard/resource-quota-card.tsx

spadgett commented on frontend/public/components/default-resource.tsx:
```diff
@@ -132,6 +136,36 @@ export const DefaultList: React.FC<TableProps & { kinds: string[] }> = (props) =
     );
   };
 
+  const TableRowForACRQ: React.FC<RowFunctionArgs<K8sResourceKind>> = ({ obj, customData }) => {
```
This doesn't belong in default-resource.tsx. Instead of changing default-resource, define a `AppliedResourceClusterQuota` list page and register it in resource-pages.ts.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,17 +71,35 @@ export const getQuotaResourceTypes = (quota) => {
+  // It has cluster-scoped data as well as namespaced data
+  // "used" is the data for the current namespace and "totalUsed" is the cluster-scoped data
+  if (isACRQ) {
+    used = _.get(quota, ['status', 'namespaces']);
```
Don't reuse the `used` variables to mean different things. Assign each temp value to a different variable.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -406,6 +500,13 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
         namespaced: false,
         optional: true,
       });
```
You shouldn't need to request both `ClusterResourceQuota` and `AppliedClusterResourceQuota`. You should only request one or the other.

rebeccaalpert replied:
@spadgett - So do you not want to show `ClusterResourceQuota` on this page? In an earlier comment, you said you always wanted to show `AppliedClusterResourceQuota.`

spadgett replied:
If this page is namespaced, it is probably better to show `AppliedClusterResourceQuota` so that you can show usage within the project.

On the `AppliedClusterResourceQuota` details page, I would link to the `ClusterResourceQuota` for users who have permission since `ClusterResourceQuota` is the editable resource.

spadgett commented on frontend/public/components/dashboard/project-dashboard/inventory-card.tsx:
```diff
@@ -163,6 +164,7 @@ export const InventoryCard = () => {
         <ProjectInventoryItem projectName={projectName} model={RouteModel} />
         <ProjectInventoryItem projectName={projectName} model={ConfigMapModel} />
         {canListSecrets && <ProjectInventoryItem projectName={projectName} model={SecretModel} />}
+        <ProjectInventoryItem projectName={projectName} model={AppliedClusterResourceQuotaModel} />
```
I would only add this if we also add `ResourceQuota`, but if we update the quota card, it's probably not needed.

rebeccaalpert commented:
/test e2e-gcp-console

rebeccaalpert commented:
This is all set @spadgett.

I defined an ACRQ list page and added ACRQs to the ResourceQuotaCard:

<img width="549" alt="Screen Shot 2021-10-04 at 2 13 08 PM" src="https://user-images.githubusercontent.com/7014965/136032517-48a39bcd-b194-4e54-a66b-6c935841f65c.png">

<img width="544" alt="Screen Shot 2021-10-04 at 2 13 15 PM" src="https://user-images.githubusercontent.com/7014965/136032551-1b4ba746-5515-40a9-b684-defe54a9eb70.png">

I updated the behavior for the ResourceQuotas list page. ClusterResourceQuotas are displayed on the ResourceQuotas list page if "all namespaces" is selected. Otherwise, ACRQs are shown. ACRQs now link to the relevant CRQ if users can list CRQs.

spadgett commented on frontend/packages/console-shared/src/components/dashboard/resource-quota-card/AppliedClusterResourceQuotaItem.tsx:
```diff
@@ -0,0 +1,47 @@
+        />
+        {scopes && <QuotaScopesInline scopes={scopes} />}
+      </div>
+      <QuotaGaugeCharts
```
I'd think we'd want to update `QuotaGuageCharts` to show both used for project and total used.

spadgett commented on frontend/packages/console-shared/src/components/dashboard/resource-quota-card/AppliedClusterResourceQuotaItem.tsx:
```diff
@@ -0,0 +1,47 @@
+  namespace,
+}) => {
+  const resourceTypes = getQuotaResourceTypes(resourceQuota);
+  const scopes = _.get(resourceQuota, 'spec.scopes');
```
Prefer optional chaining for new code.

```suggestion
  const scopes = resourceQuota?.spec?.scopes;
```

spadgett commented on frontend/packages/console-shared/src/components/dashboard/resource-quota-card/AppliedClusterResourceQuotaItem.tsx:
```diff
@@ -0,0 +1,47 @@
+export default AppliedClusterResourceQuotaItem;
+
+type AppliedClusterResourceQuotaItemProps = {
+  resourceQuota: K8sResourceKind;
```
Add a type for `AppliedClusterResourceQuota`.

spadgett commented on frontend/public/components/dashboard/project-dashboard/resource-quota-card.tsx:
```diff
@@ -21,6 +23,13 @@ const getResourceQuota = (namespace: string) => ({
   prop: 'resourceQuotas',
 });
 
+const getAppliedClusterResourceQuota = (namespace: string) => ({
```
```suggestion
const getAppliedClusterResourceQuota = (namespace: string): FirehoseResource => ({
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -42,10 +47,20 @@ const clusterResourceQuotaMenuActions = [
+const clusterQuotaReference = referenceForModel(ClusterResourceQuotaModel);
+const appliedClusterQuotaReference = referenceForModel(AppliedClusterResourceQuotaModel);
+
+const quotaKind = (quota) => {
```
Not new, but `quotaKind` is a misleading name. Is there a reason we can't just use `referenceFor`?

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,17 +72,35 @@ export const getQuotaResourceTypes = (quota) => {
 };
 
-const getResourceUsage = (quota, resourceType) => {
+const getResourceUsage = (quota, resourceType, namespace, isACRQ = false) => {
```
We shouldn't need to pass `isACRQ` around in all these functions if we're passing the `quota` object. You can simply check the obj kind and avoid the extra parameter everywhere.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -253,22 +322,50 @@ export const hasComputeResources = (resourceTypes) => {
-  const text = isClusterQuota(rq)
-    ? t('public~ClusterResourceQuota details')
-    : t('public~ResourceQuota details');
+  const kind = quotaKind(rq);
```
This function name and variable name are misleading. This is not a kind, it's a reference, which is something different.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -453,10 +586,43 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
-    menuActions={resourceQuotaMenuActions}
-    pages={[navFactory.details(Details), navFactory.editYaml()]}
-  />
+export const AppliedClusterResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
```
We're connecting to flags here, but I don't see how the flag is used. If you're already on the AppliedClusterResourceQuota page, it should only be OpenShift.

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -453,10 +586,43 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
+
+    return (
+      <MultiListPage
+        ListComponent={ResourceQuotasList}
```
We want a specific list component for AppliedClusterResourceQuota IMO that has different table columns.

spadgett replied:
I think the label selector, target project annotations, and creation date would be good (for both AppliedClusterRresourceQuota and ClusterResourceQuota).

rebeccaalpert commented:
Added tests.

rebeccaalpert commented:
Just pushing squashed commits - no changes.

spadgett commented on frontend/packages/console-shared/src/components/dashboard/resource-quota-card/AppliedClusterResourceQuotaItem.tsx:
```diff
@@ -0,0 +1,58 @@
+        />
+        {scopes && <QuotaScopesInline scopes={scopes} />}
+      </div>
+      <Text component={TextVariants.h4}>{t('public~For project')}</Text>
```
It would be better to do this in one donut chart with multiple different colored arcs (usage for current project and usage for other projects).

rebeccaalpert replied:
I wrote to Chris to see if we can get a design done for this.

cshinn replied:
Do you mean something like this? https://www.sketch.com/s/a2668252-07fe-4472-a96f-d3bf94423959/a/jgpErYb (Sorry, some of the formatting has been messed up)

spadgett replied:
I'd definitely like to hear what @cshinn thinks, but this could probably be a simple donut chart with an arc for the current project and an arc for other projects. That's what we did in OpenShift 3.x.

rebeccaalpert replied:
It looks like PatternFly won't let you do gauge charts like that design @cshinn. ChartDonutUtilization only lets you pass in one x and y coordinate. It looks like we can do a simple donut chart like @spadgett is suggesting (it looks like those let you pass in more data), but you can potentially wind up with charts that look like this (where both the cluster and namespace data are at 25%):
<img width="239" alt="Screen Shot 2021-11-05 at 8 06 05 PM" src="https://user-images.githubusercontent.com/7014965/140590824-782a534f-fa9b-4353-9ab3-bf0b76d6b004.png">

Are you cool with this or do you want to do something else?

cshinn replied:
hmm, I think that could be misleading. It's kinda kludgy, but would it be possible to make a "fake" segment that represented the unused portion?

cshinn replied:
if that's not possible, I guess it's probably best to stick with the separate charts for the project and total usage and make a feature request against PatternFly about being able to have multiple series in a utilization chart

rebeccaalpert replied:
@spadgett - I spoke to @cshinn on Slack and he said it seems like the best case for now would be to use separate charts like we had before, but maybe it’s worth making a feature request against PF. Are you ok with that?

spadgett commented on frontend/public/components/dashboard/project-dashboard/resource-quota-card.tsx:
```diff
@@ -29,10 +38,24 @@ export const ResourceQuotaCard = withDashboardResources(
     const quotas = _.get(resources.resourceQuotas, 'data', []) as FirehoseResult['data'];
-    const loaded = _.get(resources.resourceQuotas, 'loaded');
-    const error = _.get(resources.resourceQuotas, 'loadError');
+    const clusterQuotas = _.get(
```
nit: for new code, prefer optional chaining

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,23 +109,52 @@ export const getQuotaResourceTypes = (quota) => {
+  if (isACRQ) {
+    if (!clusterOnly) {
+      const allNamespaceData = _.get(quota, ['status', 'namespaces']);
+      if (namespace !== undefined) {
```
If there isn't a specific reason we need to differentiate between `undefined` and `''` or `null`, use

```suggestion
      if (namespace) {
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -401,11 +688,21 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
-        namespaced: false,
-        optional: true,
-      });
+      if (namespace === undefined) {
```
```suggestion
      if (!namespace) {
```

spadgett commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -453,10 +750,69 @@ export const ResourceQuotasPage = connectToFlags(FLAGS.OPENSHIFT)(
+  return (
+    <DetailsPage
+      {...props}
+      menuActions={[
```
It would be good to avoid duplicating the actions list for the list page and details page.

TheRealJon reviewed (changes requested):
Found a few things to address.

rebeccaalpert commented on frontend/public/module/k8s/types.ts:
```diff
@@ -319,6 +319,21 @@ export type DeploymentKind = {
+  status?: {
+    namespaces?: {
+      namespace: string;
+      status: { used: { [key: string]: number }; total: { [key: string]: number } };
```
The correct API for ACRQs is total.

rebeccaalpert commented on frontend/__tests__/components/resource-quota.spec.tsx:
```diff
@@ -1,6 +1,56 @@
+        },
+        {
+          namespace: 'test-namespace2',
+          status: { used: { 'limits.cpu': '1' }, total: { 'limits.cpu': '2' } },
```
The correct API for ACRQs is total.

rebeccaalpert commented on frontend/__tests__/components/resource-quota.spec.tsx:
```diff
@@ -1,6 +1,56 @@
+      namespaces: [
+        {
+          namespace: 'test-namespace',
+          status: { used: { 'limits.cpu': '0' }, total: { 'limits.cpu': '2' } },
```
The correct API for ACRQs is total.

rebeccaalpert commented on frontend/__tests__/components/resource-quota.spec.tsx:
```diff
@@ -98,3 +148,49 @@ describe('Check cluster quota table columns by ResourceUsageRow', () => {
+        },
+        {
+          namespace: 'test-namespace2',
+          status: { used: { 'limits.cpu': 1 }, total: { 'limits.cpu': 2 } },
```
The correct API for ACRQs is total.

rebeccaalpert commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -365,18 +588,83 @@ export const ResourceQuotasList = (props) => {
     <Table
       {...props}
-      aria-label={ResourceQuotaModel.labelPlural}
+      aria-label={t('public~ResourceQuotas')}
```
I'm aware of these (I am the one who added them), but Sam's told me in the past to just write out the model name.

rebeccaalpert commented:
Updated charts and addressed feedback.

rebeccaalpert commented:
/test e2e-gcp-console

rebeccaalpert commented:
Addressed test failures. The remaining e2e test failure appears to be a known flake and Christoph is currently fixing it.

TheRealJon reviewed (changes requested):
A couple of minor things and one question. Is it expected that the ACRQ list page shows no results when "All Projects" is selected?

rebeccaalpert commented on frontend/public/components/resource-quota.jsx:
```diff
@@ -57,23 +101,68 @@ export const getQuotaResourceTypes = (quota) => {
   const specPath = isCluster ? ['spec', 'quota', 'hard'] : ['spec', 'hard'];
   const usedPath = isCluster ? ['status', 'total', 'used'] : ['status', 'used'];
+  const used = _.get(quota, [...usedPath, resourceType]);
+
   const max =
     _.get(quota, [...statusPath, resourceType]) || _.get(quota, [...specPath, resourceType]);
-  const used = _.get(quota, [...usedPath, resourceType]);
   const percent = !max || !used ? 0 : (convertToBaseValue(used) / convertToBaseValue(max)) * 100;
+
```
Good call; I used to use this code, but split out the ACRQ stuff in the end.

jhadvig commented on frontend/public/module/k8s/types.ts:
```diff
@@ -326,6 +326,20 @@ export type DeploymentKind = {
+export type AppliedClusterResourceQuotaKind = {
+  spec?: {
+    selector?: MatchLabels | Selector;
+    quota?: { hard: { [key: string]: number }; scopes?: string[] };
```
`quota` is missing `ScopeSelector` field, check
https://github.com/openshift/api/blob/be1be0e89115702f8b508d351c4f5c9a16e5ae95/quota/v1/types.go#L37
and 
https://github.com/openshift/api/blob/be1be0e89115702f8b508d351c4f5c9a16e5ae95/vendor/k8s.io/api/core/v1/types.go#L5749

rebeccaalpert commented:
Yeah, I think it's expected that the ACRQ list page shows no results when "All Projects" is selected, but I'm not an expert on this. It needs a namespace for the API I suspect we're using. [There is an api for showing all ACRQs](https://docs.openshift.com/container-platform/4.9/rest_api/schedule_and_quota_apis/appliedclusterresourcequota-quota-openshift-io-v1.html), but I'm not sure how that ties in to our existing code. Maybe someone who's more familiar with that could take a look as a follow-on project.

TheRealJon commented on frontend/public/module/k8s/types.ts:
```diff
@@ -326,6 +326,26 @@ export type DeploymentKind = {
 
+export type AppliedClusterResourceQuotaKind = {
+  spec?: {
+    selector?: MatchLabels | Selector;
```
This doesn't match the API:
```suggestion
    selector?: {
      labels?: Selector;
      annotations?: MatchLabels;
    };
```

TheRealJon reviewed (changes requested):
One more tweak to the ACRQ type definition

TheRealJon reviewed (approved):
/lgtm

rebeccaalpert commented:
/assign @ahardin-rh @opayne1
for docs approval

/assign @yapei
for QE approval

/assign @RickJWagner
for px approval

rebeccaalpert commented:
Putting on hold because the merge queue is blocked due to a test flake. This is ready for docs, product, and QE review though.

RickJWagner commented:
/label px-approved

opayne1 commented:
/label docs-approved

opayne1 commented:
/label docs-approved

schituku commented:
/lgtm

schituku commented:
/label qe-approved
